### PR TITLE
Block Directory: Throw error if we have an issue registering blocks.

### PR DESCRIPTION
**Pull request by StevenDufresne**
## Description
When a third party block fails during the installation process, we should alert the user to give them context.

## How has this been tested?
1.  Search for "Waves" -> At the time of the PR, that block is currently breaking (but probably fixed by now)
2. Click 'Add Block' 
3. Expect to see an error messages below:

## Types of changes
1. Update the structure of Error Notice state.
2. Change button to`reload` from `retry` if error is known to be `fatal`.
3. Specifically handle API errors `folder_exists`, `unable_to_connect_to_filesystem`.

## Assumptions
This PR doesn't try to handle all the use cases but instead identifies the ones we know to be fatal or introduce and awkward experience. All other unhandled but thrown errors will allow user's the opportunity to retry.

## Screenshots
| Already Installed  |  Failed Install | Failed Registration |
| ------------- | ------------- | ------------- |
|  Error msg  | Error msg |   Error msg |
|  | Note: Probably should add more error code handling. Currently only `unable_to_connect_to_filesystem` |  |

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by tellyworth** [comment]
> I wonder if we should maybe create a fatal error state that doesn't allow retries?

I think so, yes. If it's a problem that we know can't be solved by retrying then that's a better experience. Ideally we shouldn't return broken blocks in search results in the first place, but it makes sense to have a way of handling them.

**Comment by StevenDufresne** [comment]
@ryelle What are your thoughts here? 

I don't want to invest too much time in this direction if we think there's a better approach (which is why this is still draft).

**Comment by ryelle** [comment]
I think you're on the right track. Maybe you could show a "reload" button instead of "retry" for these kinds of errors? If it's not a hassle - ideally this situation would be very rare, since the wp.org block validation _should_ catch these errors before the blocks get added to the directory, right?

**Comment by ryelle** [comment]
I added 2a28f968daced4ad3e7f5fda6bbc52375638d197 to disable the "Add Block" button when a block has a fatal notice, and ad5e50c1c43274a1485291d8522d3b5ea5609274 to clean up the error messages a bit.

**Comment by ryelle** [comment]
It also looks like the tests need to be updated, or the `clearErrorNotice` should go back to setting `notice` to false. Removing that item from the notice state would probably be a cleaner solution, if you do want to change `clearErrorNotice`.

File: packages/block-directory/src/components/downloadable-block-notice/style.scss [hunk 1]
```diff
@@ -4,5 +4,8 @@
 
 .block-directory-downloadable-block-notice__content {
 	padding-right: 12px;
-	margin-bottom: 8px;
+}
+
+.block-directory-downloadable-block-notice__content + * {
+	margin-top: 8px;
```
[the remark this answers is not in the document]
Comment (StevenDufresne) [reply packages/block-directory/src/components/downloadable-block-notice/style.scss:10]: Ah good catch. This was from the iteration where the button didn't exist if fatal. Now we always persist a button (reload btn now), so I'll revert back.

File: packages/block-directory/src/store/actions.js [hunk 2]
```diff
@@ -166,15 +192,19 @@ export function setIsInstalling( blockId, isInstalling ) {
...
  *
  * @param {string} blockId The ID of the block plugin. eg: my-block
- * @param {string} notice  The message shown in the notice.
+ * @param {string} msg  The message shown in the notice.
+ * @param {boolean} isFatal Whether the user can recover from the error
  *
  * @return {Object} Action object.
  */
-export function setErrorNotice( blockId, notice ) {
+export function setErrorNotice( blockId, msg, isFatal = false ) {
```
[the remark this answers is not in the document]
Comment (StevenDufresne) [reply packages/block-directory/src/store/actions.js:200]: Yep, sounds good.

**Comment by StevenDufresne** [comment]
> It also looks like the tests need to be updated, or the `clearErrorNotice` should go back to setting `notice` to false. Removing that item from the notice state would probably be a cleaner solution, if you do want to change `clearErrorNotice`.

Yep, will get to those today.

**Review by ryelle — APPROVED** [review]
👍  Fixes the issue with errors we can detect, so this is a good improvement.

**Comment by TimothyBJacobs** [comment]
Does it make sense to do this for all `500` status code errors? Those should theoretically already denote "fatal" errors, or am I misunderstanding what you mean by fatal in this case.

**Comment by StevenDufresne** [comment]
> Does it make sense to do this for all `500` status code errors? Those should theoretically already denote "fatal" errors, or am I misunderstanding what you mean by fatal in this case.

No, you are correct. Under a bit of a time crunch, we focused on handling the error cases we saw during third party testing to provide more useful user error messages. We'll follow this up and test more api errors, and most likely cover all `500` cases with 'reload' type error messaging where applicable.

**Pull request merged**